src/CLICommand/action.ts

```diff
@@ -0,0 +1,120 @@
+import { ActionReturnType, AsyncActionCallback, ParsedParameters } from './cli';
+import { ERROR_EXIT_CODE } from './error_codes';
+
+/**
+ * A wrapper for the command action callback
+ */
+export class CLIAction {
+	private _promiseInstance?: Promise<ActionReturnType>;
+	private _parsedParameters: ParsedParameters = {};
+	private awaiterDone?: (value?: ParsedParameters, error?: Error) => void;
+	private static _runningAction?: CLIAction;
+
+	/**
+	 * Create an instance of CLIAction
+	 * @param {AsyncActionCallback} actionCallback the handler method of the action
+	 */
+	constructor(private readonly actionCallback: AsyncActionCallback) {}
+
+	private get promiseInstance(): Promise<ActionReturnType> {
+		if (!this._promiseInstance) {
```

Comment:
This pattern is not my favorite. I'd rather do type design to make illegal states totally unrepresentable. But this will be ok for now.